### PR TITLE
Issue #474 End statements should not be used

### LINKED ISSUE #474: End statements should not be used
RSPEC: https://jira.sonarsource.com/browse/RSPEC-1147

### PR DESCRIPTION
**Pull request by tvsonar**
Fixes #474

### CODE REVIEW

File: src/SonarLint/Rules.Description/S1147.html [hunk 1]
```diff
@@ -0,0 +1,24 @@
+<p>
+    Calling <code>ToString()</code> on an object should always return a string. Returning <code>null</code> instead contravenes the method's implicit contract.
```
Comment (dbolkensteyn) [inline src/SonarLint/Rules.Description/S1147.html]: @tvsonar euh? seems to be the wrong description

File: src/SonarLint/Rules/EndStatementUsage.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+{
+    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+    [SqaleConstantRemediation("30min")]
+    [SqaleSubCharacteristic(SqaleSubCharacteristic.SecurityFeatures)]
+    [Rule(DiagnosticId, RuleSeverity, Title, IsActivatedByDefault)]
+    [Tags(Tag.Cwe, Tag.Pitfall)]
+    public class EndStatementUsage : DiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S1147";
+        internal const string Title = "\"End\" statements should not be used";
```
Comment (dbolkensteyn) [inline src/SonarLint/Rules/EndStatementUsage.cs]: @tvsonar why not also mention "stop" in the title as it also gets detected by this rule?

Comment (dbolkensteyn) [inline src/SonarLint/Rules/EndStatementUsage.cs]: @tvsonar I wouldn't merge this rule with the Stop one afterall. Currently the existing rule in the VB.NET plugin only detects the usage of the `End` statement, which is a kind of maintainability issue, whereas it looks like `Stop` is not expected to be used in production code as it that one indeed fails to free variables and stuff

Comment (dbolkensteyn) [inline src/SonarLint/Rules/EndStatementUsage.cs]: @tvsonar btw it looks like `Stop` doesn't interrupt the control flow, it sends a signal to the debugger:
```
Stop

Console.WriteLine("Hello, world!")
Console.ReadLine()
```

The `Stop` statement acts as a breakpoint. When clicking Resume on the debugger, the program just continues as if the `Stop` statement wasn't there at all. When launched outside of Visual Studio, the `Stop` statement leads to a "crash" prompt, which if you close it, simply continues to execute the program. That happens in both "Debug" and "Release" configuration. See https://msdn.microsoft.com/en-us/library/6acc062s.aspx

=> I don't expect to see any usage of the `Stop` statement in production code, it probably is a very bad replacement to breakpoints

My feeling is that `Stop` could deserve another rule, with a critical severity and a dedicated explanation mentioning that breakpoints are a better alternative.

Comment (tvsonar) [inline src/SonarLint/Rules/EndStatementUsage.cs]: The Stop statement suspends execution, but unlike End, it does not close any files or clear any variables, unless it is encountered in a compiled executable (.exe) file. In https://msdn.microsoft.com/en-us/library/0wt87xba.aspx

Comment (tvsonar) [inline src/SonarLint/Rules/EndStatementUsage.cs]: Created #551 for the Stop statement

**Pull request merged**